### PR TITLE
Issue/#637 mq service initialization race condition

### LINKED ISSUE #637: Race condition in MQService initialization
Looks like there is a pesky concurrency issue in the message service:

```bash
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 587, in run_until_complete
    return future.result()
  File "server.py", line 73, in main
    service.initialize() for service in services.values()
  File "/code/server/game_service.py", line 63, in initialize
    await self._message_queue_service.declare_exchange("faf-rabbitmq")
  File "/code/server/message_queue_service.py", line 80, in declare_exchange
    new_exchange = await self._channel.declare_exchange(
AttributeError: 'NoneType' object has no attribute 'declare_exchange'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aiormq/connection.py", line 374, in __reader
    weight, channel, frame = await self.__receive_frame()
  File "/usr/local/lib/python3.7/site-packages/aiormq/connection.py", line 326, in __receive_frame
    frame_header = await self.reader.readexactly(1)
  File "/usr/local/lib/python3.7/asyncio/streams.py", line 679, in readexactly
    await self._wait_for_data('readexactly')
  File "/usr/local/lib/python3.7/asyncio/streams.py", line 473, in _wait_for_data
    await self._waiter
concurrent.futures._base.CancelledError
```

1. Game service initialize and message queue service initialize are both started and awaited with `asyncio.gather`
2. message queue service successfully connects and yields on `_connection.channel`
3. Game service calls `declare_exchange`

Now the message queue service is in a state where `_connection` is initialized, but `_channel` is still None

### PR DESCRIPTION
**Pull request by cleborys**
Adds `_is_ready` to `MessageQueueService` keeping track of whether a connection is fully established (i.e. connection *and* channel). Methods that touch ths are `initialize()`, `shutdown()`, and `reconnect()`.
Since spamming e.g. reconnect could potentially leave things in a bad state, I also added an `_initialization_lock` and `_is_ready` is only modified by methods holding the lock.

Closes #637

### CODE REVIEW

File: server/message_queue_service.py [hunk 1]
```diff
@@ -23,6 +23,8 @@ def __init__(self) -> None:
         self._channel = None
         self._exchanges = {}
         self._exchange_types = {}
+        self._is_ready = False
+        self._initialization_lock = asyncio.Lock()
```
Comment (Askaholic) [inline server/message_queue_service.py:27]: Btw I wrote a decorator for this here https://github.com/FAForever/server/pull/629/commits/cda403335f72b7ca428ef6cb553c8bbc1c963193. Maybe I should extract it to a separate PR?

It lets you do:
```python3
class SomeClass:

    @synchronizedmethod
    async def some_func(self, args):
        ... 
```
to wrap a method with one lock per instance, or 
 
```python3
@synchronized
async def some_func(args):
    ...
```
to lock a function with a global lock (works for methods too, but you will only be able to call the method on one instance at a time).

File: server/message_queue_service.py [hunk 2]
```diff
@@ -31,12 +33,14 @@ def __init__(self) -> None:
...
 
     async def initialize(self) -> None:
-        if self._connection is not None:
-            return
+        async with self._initialization_lock:
+            if self._connection is not None:
+                return
 
-        await self._connect()
+            self._is_ready = await self._connect()
```
Comment (Askaholic) [inline server/message_queue_service.py:40]: why not make `_connect` set the attribute directly?

File: server/message_queue_service.py (hunk 1 above)
Comment (Askaholic) [reply server/message_queue_service.py:27]: Oh I guess you are using the lock in more than one function so maybe it's not useful here.

Comment (cleborys) [reply server/message_queue_service.py:27]: I guess one could add something like `@synchronized_group(lock_name)` managing a global dictionary of locks that can be shared between functions/methods, if you want.

File: server/message_queue_service.py (hunk 2 above)
Comment (cleborys) [reply server/message_queue_service.py:40]: Because I also want to call `_connect` from `reconnect` and there I don't want to set the flag until after the previous exchanges have been re-declared.

Comment (Askaholic) [reply server/message_queue_service.py:40]: Hmm, maybe it would be better to propagate exceptions then. I feel like passing around a "
success" flag as return value is pretty clunky. Gives me nightmares about C.

File: server/message_queue_service.py (hunk 1 above)
Comment (Askaholic) [reply server/message_queue_service.py:27]: I worked on it a bit more, and I came up with that same solution, except I just made it an optional parameter for the existing decorator. So it would be:

```python3
class Foo:
    def __init__(self):
        self._the_lock = asyncio.Lock()

    @synchronizedmethod("_the_lock")
    async def bar(self):
        ...

    @synchronizedmethod(lock="_the_lock")
    async def baz(self):
        ...

```

File: server/message_queue_service.py (hunk 2 above)
Comment (cleborys) [reply server/message_queue_service.py:40]: Sure thing.

File: server/message_queue_service.py [hunk 3]
```diff
@@ -11,6 +11,10 @@
 from .decorators import with_logger
 
 
+class ConnectionAttemptFailed(Exception):
```
Comment (Askaholic) [inline server/message_queue_service.py:14]: Maybe inherit from ConnectionError?

```suggestion
class ConnectionAttemptFailed(ConnectionError):
```

File: server/message_queue_service.py [hunk 4]
```diff
@@ -31,12 +37,18 @@ def __init__(self) -> None:
...
 
-        await self._connect()
+            try:
+                await self._connect()
+            except ConnectionAttemptFailed:
+                return
+            self._is_ready = True
 
     async def _connect(self) -> None:
+        """ Returns True on success. """
```
Comment (Askaholic) [inline server/message_queue_service.py:51]: Nope
```suggestion
```

File: server/message_queue_service.py [hunk 5]
```diff
@@ -49,29 +61,30 @@ async def _connect(self) -> None:
...
+            )
+            raise ConnectionAttemptFailed
         except ProbableAuthenticationError:
             self._logger.warning(
-                "Unable to connect to RabbitMQ. Incorrect credentials?",
-                exc_info=True
+                "Unable to connect to RabbitMQ. Incorrect credentials?", exc_info=True
             )
-            return
+            raise ConnectionAttemptFailed
```
Comment (Askaholic) [inline server/message_queue_service.py:72]: Maybe use the `raise from` syntax.

```suggestion
            raise ConnectionAttemptFailed from e
```
(need to change the `except` block to name the exception)

File: server/message_queue_service.py [hunk 6]
```diff
@@ -31,12 +37,18 @@ def __init__(self) -> None:
         config.register_callback("MQ_PORT", self.reconnect)
 
     async def initialize(self) -> None:
-        if self._connection is not None:
-            return
+        async with self._initialization_lock:
+            if self._connection is not None:
```
Comment (Askaholic) [inline server/message_queue_service.py:41]: Should we be checking `is_ready` here instead?

Comment (cleborys) [reply server/message_queue_service.py:41]: Yes that seems better. TBH one could probably also replace the flag by `self._connection is not None` checks everywhere else, but having the flag feels more readable.

**Comment by Askaholic** [comment]
Do we need to update something in GameService too to wait for the MQService to be ready before calling `declare_exchange`?

**Comment by cleborys** [comment]
Hm... Instead I chose to have `declare_exchange` call the initialization if necessary and added a test to make sure there is only one initialization, even if you call it several times.
(The public `declare_exchange` has a check whether initialization is necessary, the private `_declare_exchange` does not, since you want to be able to re-declare all old exchanges when reconnecting, before you mark the service as ready again.)

Coincidentally, I assume the `asyncio.Lock` is not like a `threading.RLock`, so calling a `@synchronizedmethod("the_lock")` from another `@synchronizedmethod("the_lock")` will deadlock?

**Comment by Askaholic** [comment]
Yes I expect it would

File: integration_tests/test_message_queue/test_message_queue_service.py [hunk 7]
```diff
@@ -144,3 +146,30 @@ async def test_incorrect_vhost(mocker, caplog):
...
+        i = 0
+        while service._channel is None:
+            await service.declare_exchange(f"exchange_{i}")
+            await asyncio.sleep(0)
+            i += 1
+
+    await asyncio.gather(keep_declaring_exchanges(), service.initialize())
+
+
+async def test_declaring_exchange_without_initialization():
```
Comment (Askaholic) [inline integration_tests/test_message_queue/test_message_queue_service.py:164]: What was the deal with these tests again? Are we not able to get them to run in travis?

**Review by Askaholic — APPROVED** [review]
Looks fine!

File: integration_tests/test_message_queue/test_message_queue_service.py (hunk 7 above)
Comment (cleborys) [reply integration_tests/test_message_queue/test_message_queue_service.py:164]: We can, but they rely on running the faf-rabbitmq container, which I believe we decided not to want to depend on.
It would be quick to add them back to travis, for example the league service is doing that:
https://github.com/cleborys/faf-league-service/blob/9bec96d5f1b0f9bdf69baff014489fc7b3482a9f/.travis.yml#L27
https://github.com/cleborys/faf-league-service/blob/develop/ci/init-rabbitmq.sh

Comment (Askaholic) [reply integration_tests/test_message_queue/test_message_queue_service.py:164]: Hmm, yea it might be nice to do that in a separate PR. I can't remember why we decided not to before. I remember we put that stuff in the integration tests because we didn't want people to need to install rabbitmq just to develop the server, but I would think on Travis it would be fine.

Comment (cleborys) [reply integration_tests/test_message_queue/test_message_queue_service.py:164]: Ah, I see. I didn't see that distinction between contributors running it and travis running it, that's why I moved it to the integration tests. Moving it to travis would also mean the docker repository won't evict `faf-rabbitmq` because of six months of inactivity :stuck_out_tongue_closed_eyes: 

By the way, "installing rabbitmq" is literally only running the script linked above which pulls the docker image from the repository and adds the appropriate user, so the only drawback I'd see for contributors is that the image takes up some space.

Comment (Askaholic) [reply integration_tests/test_message_queue/test_message_queue_service.py:164]: Hmm, yea that's not a big deal.
